Update CONTRIBUTING file to match current practices.

dwlehman opened the pull request:
You can see a preview of what it will look like [here](https://gist.github.com/dwlehman/ceaf09d7ae1ac1c9e13da79285f5f738).

vpodzime commented on CONTRIBUTING:
````diff
@@ -1,282 +1,113 @@
+      ```
+      Take a look to make sure things are as expected, then...
+      ```
+      git push
````
Does this actually work? Maybe I have some different config, but just ``git push`` pushes only the current branch to its upstream. What I have to do is something like this ``git push origin master 2.0-devel 2.1-devel 3.0-devel``.

dwlehman replied:
That is an example of merging 2.0-devel to 2.1-devel. Maybe I should make that more clear.

vpodzime commented:
Apart from the above question I'd like to ask why this document uses the markdown syntax while we use the ReST syntax everywhere else (AFAICT).

dwlehman commented:
I used markdown because I forgot that github will render the ReST nicely as well when people click the link when creating a pull request. I'll convert it.

dwlehman commented:
I converted it to ReST and made it clear that that example was merging `2.0-devel` into `2.1-devel`.

I also added a note about release branches since it seems we're bringing them back.

dwlehman commented:
Also updated demo version [here](https://gist.github.com/dwlehman/ceaf09d7ae1ac1c9e13da79285f5f738) (same link as before).

vpodzime commented:
Looks good to me. And since it is now ReST, I think it would be nice to add a table of contents to the beginning of the file, perhaps with depth=1?
